### PR TITLE
1516 program permissions

### LINKED ISSUE #1516: Remove fine grain access control for document types
Currently document access is restricted by document type. However, it's enough to restrict document access by program.

- [x] redefine dynamic document permission to only include program types as context
- [x] update permission setup in init scripts (remove non-program document types)
- [ ] ~Calculate the allowed docs list from the `document_registry` from the list of program names from the program permissions~
- [x] rename `CapabilityTag::DocumentType` ...
- [x] handle access for patient documents since they are not belonging to a special program

### PR DESCRIPTION
**Pull request by clemens-msupply**
Fixes #1516 

# 👩🏻‍💻 What does this PR do? 

Changes the permission checks to just check for context rather than the document type. This required to store the context in all relevant tables. The context is currently just the programs document name but can, in a future PR, point to an entry in the permission table.

# 🧪 How has/should this change been tested? 

This change requires an updated 4d server. The required changes are already in the mSupply develop branch.
To initialize the data please use the `1516-program-permissions` branch from the programs schema repository.

### CODE REVIEW

File: client/packages/common/src/types/schema.ts [hunk 1]
```diff
@@ -800,38 +800,33 @@ export type DocumentRegistryConnector = {
...
   children: Array<DocumentRegistryNode>;
-  context: DocumentRegistryNodeContext;
+  documentContext: Scalars['String'];
   documentType: Scalars['String'];
   formSchemaId: Scalars['String'];
   id: Scalars['String'];
   jsonSchema: Scalars['JSON'];
   name?: Maybe<Scalars['String']>;
   parentId?: Maybe<Scalars['String']>;
+  type: DocumentRegistryTypeNode;
```
Comment (clemens-msupply) [inline client/packages/common/src/types/schema.ts:820]: This was the old `context` which is now the type of the document registry entry, e.g. Patient | Encounter | ProgramEnrolment

File: client/packages/programs/src/api/api.ts [hunk 2]
```diff
@@ -196,32 +196,6 @@ export const getEncounterQueries = (sdk: Sdk, storeId: string) => ({
...
-        filter: { context: { equalTo: context } },
-        storeId,
-      });
-      const entries = result?.documentRegistries;
-
-      if (entries?.__typename === 'DocumentRegistryConnector') {
-        return entries.nodes;
-      }
-      throw new Error('Error querying document registry by context');
-    },
```
Comment (clemens-msupply) [inline client/packages/programs/src/api/api.ts:224]: I found working with the more generic documentRegistries query was much easier than having these specialized queries...

**Review by andreievg — COMMENTED** [review]
In general looks solid (can't see any bugs). I have a lot of comments and questions, this is the first time i actually got to review program code (so a lot of comments that are not related to this PR, but I wanted to understand some functionality that is being touched to give this a thorough review)

Is there general documentation somewhere describing the schema, schema relations, logic that is tied to configurations (especially constants in configurations that have logic attached), program events etc.. ? It would help tremendously for reviewer. I ended up looking at init scripts and quite a bit of code, to try and understand how schema would work

Apart from the questions/comments in the review:

The change from program to context seems quite generic, it may apply to document but it seem to obfuscate the program relationship for other tables (I though they have an intrinsic relationship to a program ? i.e. concept of enrolement, encounter are program based) I remember originally I was in favour of program_id on user_permissions table, since that is the only use case we currently have, but we couldn't agree on it, then a compromise was made on a separate table, looks like it didn't end up happening and it just resolved back to your original design. Maybe the generic solution is fine if it's backed up by solid documentation (something that can be read before jumping at review or work on this functionality that explains it very well). 

Also documents have `owner_name_id` logic, do you still consider them generic (would we have a lot of optionals in document if it's used for other record types )?

Is there a reason why there is so much de-normalisation in the schema, like a document.form_schema_id, document.r#type, document.context, vs just document.document_registry_id, similar with encounter, encounter.document_type, encounter.context (can come from document or related document_registry), also document.name looks like a concatenation still rather then using owner_name_id patient link that exists ?  

Is update document graphql needed ? (don't see it used anywhere) 

We've been chatting about how to streamline review process, and added a couple of things [PRs for Reviewer](https://github.com/openmsupply/open-msupply/wiki/Code-Review-Comments---General#remember-prs-are-for-the-reviewer) and [Issue targeted changes](https://github.com/openmsupply/open-msupply/wiki/Code-Review-Comments---General#should-i-make-small-unrelated-fixes-in-my-pr). Although I didn't see many changes related to the later (althought the small validation changes could be done in a new PR, since this PR is already quite large), I think the former could have improved by:
* Doing a couple of stacked PRs (i.e. schema and service changes, and front end and graphql changes, or by document type) 
* Explaining the changes a bit more, i.e. "for enrolement, removed program and added document_type and context, document_type is needed because context is not availble to front end and document_type was used to specify which program enrolement is for", "added an extra check for encounter (that associated enrolement exists) and added check to patient edit to check that document registry already exists"
* Schema diagram of how permissions related to documents and their Derivatives (enrolements, patient, encounters)
* Example of configuration change between pervious and current way

I only had a small look at front end code (would assume type safety with codegen would guarantee that changes are safe, and would assume you've tested all affected areas)

## Tests

This will become more important now that we will have testers, but even without them, I think a bit more details about how and what to test, with a few more use cases: 

"To get dispensing going with program: 1 - Set a store to be disepnsary, 2 - run 'yarn build' from root of programschemas repository then run 'npx ts-node generate_4d_init_script.ts' from 4dinit folder, use generated .4dm file and insert into 4d method (this can be run via footrunner code). Adjust array of users in the init script to include one of the users in data file"   

Above could be generic instruction that you can link in a PR for testers or devs

And then list some instructions (script modifications) for user cases beyond 'all permissions for all users for all stores'

File: server/graphql/programs/src/mutations/program_enrolment/insert.rs [hunk 3]
```diff
@@ -39,7 +39,7 @@ pub fn insert_program_enrolment(
             store_id: Some(store_id.clone()),
         },
     )?;
-    let allowed_docs = user.capabilities(CapabilityTag::DocumentType);
+    let allowed_ctx = user.capabilities(CapabilityTag::ContextType);
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply server/graphql/programs/src/mutations/program_enrolment/insert.rs:42]: do you mean why CapabilityTag? I think we can remove it when having a context table. Just briefly what the motivation was: a permission context comes with a particular permission. If the context values are not unique there could be the problem that you get a context from the wrong permission and then you get the wrong access. For this reason there was the option to tag the context to keep them apart and group them... However, when referring to a particular row in a context table this wouldn't be a problem anymore.

File: server/service/src/programs/patient/upsert.rs [hunk 4]
```diff
@@ -171,17 +174,34 @@ fn patient_belongs_to_store(
...
 }
 
+fn validate_document_type(
+    ctx: &ServiceContext,
+    document_type: &str,
+) -> Result<Option<DocumentRegistry>, RepositoryError> {
+    let mut entry = DocumentRegistryRepository::new(&ctx.connection).query_by_filter(
+        DocumentRegistryFilter::new()
+            .r#type(DocumentRegistryType::Patient.equal_to())
+            .document_type(EqualFilter::equal_to(document_type)),
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply server/service/src/programs/patient/upsert.rs]: Yes by convention the is only one "Patient" type right now. I removed it.

File: server/service/src/programs/encounter/update.rs
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply server/service/src/programs/encounter/update.rs:1]: yes we could enforce that new documents match the latest version of the document_registry. At the moment its only checked that each document has the correct schema.

File: server/service/src/programs/encounter/insert.rs [hunk 5]
```diff
@@ -238,6 +267,47 @@ mod test {
             .upsert_one(&schema)
             .unwrap();
 
+        let program_type = "ProgramType".to_string();
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply server/service/src/programs/encounter/insert.rs]: It's the document type for the program enrolment, I renamed it to make it more clear.

File: server/service/src/programs/encounter/insert.rs [hunk 6]
```diff
@@ -141,33 +142,60 @@ fn generate(
...
+
 fn validate_patient_program_exists(
     ctx: &ServiceContext,
     patient_id: &str,
-    program: &str,
-) -> Result<bool, RepositoryError> {
-    let doc_name = patient_doc_name(patient_id, program);
+    encounter_registry: DocumentRegistry,
+) -> Result<Option<Document>, RepositoryError> {
+    let Some(enrolment_registry_id) = encounter_registry.parent_id else {
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply server/service/src/programs/encounter/insert.rs:167]: I was thinking about his one as well while doing this PR. It refers to the parent DocumentRegistry entry. For example, encounter has a parent program enrolment. This was a simple way to make these relations. However, the way this has evolved it isn't really needed. We now have the context in this table as well, and we can get the parent entry by looking up the enrolment with the same context...

File: server/repository/src/db_diesel/program_enrolment_row.rs [hunk 7]
```diff
@@ -13,8 +13,9 @@ use serde::{Deserialize, Serialize};
 table! {
     program_enrolment (id) {
         id -> Text,
-        program -> Text,
+        document_type -> Text,
         document_name -> Text,
+        context -> Text,
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply server/repository/src/db_diesel/program_enrolment_row.rs:18]: Its the same here but I think just context is nicer. The event table uses document_context though, will fix this to make it consistent.

**Comment by clemens-msupply** [comment]
thanks for the feedback!

> Is there general documentation somewhere describing the schema, schema relations, logic that is tied to configurations (especially constants in configurations that have logic attached), program events etc.. ? It would help tremendously for reviewer. I ended up looking at init scripts and quite a bit of code, to try and understand how schema would work
>

No not really, but yes writing up the documentation would be a good think to do soon!

> The change from program to context seems quite generic, it may apply to document but it seem to obfuscate the program relationship for other tables (I though they have an intrinsic relationship to a program ? i.e. concept of enrolement, encounter are program based) I remember originally I was in favour of program_id on user_permissions table, since that is the only use case we currently have, but we couldn't agree on it, then a compromise was made on a separate table, looks like it didn't end up happening and it just resolved back to your original design. Maybe the generic solution is fine if it's backed up by solid documentation (something that can be read before jumping at review or work on this functionality that explains it very well).

I mainly needed the context for permission checks here. But I think it would be good to refer to the programs table, e.g. in encounter and program_enrolment tables once needed. There is still the plan for a context table or what design do you mean? (at least this isn't my original design :) ). Lets catchup about it!

> Also documents have `owner_name_id` logic, do you still consider them generic (would we have a lot of optionals in document if it's used for other record types )?

I like the analogy for document to files. A Unix file has an owner and a group (context). Don't see a problem of having a null `owner_name_id` if a document is used for something that isn't attached for a name.

> Is there a reason why there is so much de-normalisation in the schema, like a document.form_schema_id, document.r#type, document.context, vs just document.document_registry_id, similar with encounter, encounter.document_type, encounter.context (can come from document or related document_registry),

The idea for the document registry was that it contains the reference to the latest schema but old documents could have used other schemas. For this reason the document records the schema ID that has been used at the point the document has been submitted. For `type` and `context` it's more convenience, these columns shouldn't change for an existing document... 

> also document.name looks like a concatenation still rather then using owner_name_id patient link that exists ?
 
For the document names, I would like to have a deterministic way to create a document identifier. For example, if a patient is enrolled in the same program at two different clinics there should only be one enrolment. This is possible by creating the name manually using some convention. For encounters a unique ID would be enough though. The name could be shortened though, now that the patient ID is stored in the document table!

> Is update document graphql needed ? (don't see it used anywhere)

Its not used at the moment, it was at some point but not anymore. I created an issue #1942

**Comment by clemens-msupply** [comment]
I just realized there wasn't an issue for the context table! I created this one:
https://github.com/openmsupply/open-msupply/issues/1944
I also put two related points from the comments in.

**Comment by clemens-msupply** [comment]
> Is there general documentation somewhere describing the schema, schema relations, logic that is tied to configurations (especially constants in configurations that have logic attached), program events etc.. ? It would help tremendously for reviewer. I ended up looking at init scripts and quite a bit of code, to try and understand how schema would work

There are actually:
`service/src/programs/Programs.md`
`service/src/programs/ProgramEvents.md`
but at least the first one needs an update...

**Review by andreievg — CHANGES_REQUESTED** [review]
Thank you for replies and changes so far.

1. I think we can use enrolment table for `validate_patient_program_exists` and remove parent_id from document_registry
2. I also think that program for enrolment and other tables was clearer. And I think it should be document_context, since it explicit in the meaning, and other columns that denormalised from document have `document_` prefix (in the future I would assume it would be changed to program_id, and program would then link to context, i.e. encounter ->  program_id -> context -> permission)

nit: document_registry or context as per this comment: https://github.com/openmsupply/open-msupply/pull/1928#discussion_r1242885118

> The idea for the document registry was that it contains the reference to the latest schema
> yes we could enforce that new documents match the latest version of the document_registry

When say encounter configuration is updated (new field is added) would you create a new document_registry or keep existing one ?

> For type and context it's more convenience, these columns shouldn't change for an existing document...

I think we need an issue to normalise and generally avoid denormalisation (unless it's absolutely necessary, like when we need to snapshot state of related record since it may change), it maybe convenient when it's originally written (when it's fully understood), but duplicates can create confusion when someone new is looking at the code or when looking at the code in the future.
 
> For example, if a patient is enrolled in the same program at two different clinics there should only be one enrolment

How would this work in practice ? Since I would assume system would not allow you to enrol patient more then once and all patient related data would sync to sites where they 'exist', would it be if patient 'exists' in both sites and one sites does not have internet while patient visits one site and enrols in a program then visit the site with no internet and enrols there ?

Thanks for pointing out the `.md` file, I'll have a read through it

File: server/service/src/programs/encounter/insert.rs [hunk 8]
```diff
@@ -46,21 +46,21 @@ pub fn insert_encounter(
...
     user_id: &str,
     input: InsertEncounter,
-    allowed_docs: Vec<String>,
+    allowed_ctx: Vec<String>,
 ) -> Result<Document, InsertEncounterError> {
     let patient = ctx
         .connection
         .transaction_sync(|_| {
-            let (encounter, clinician) = validate(ctx, &input)?;
+            let (encounter, program_enrolment, clinician) = validate(ctx, &input)?;
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply server/service/src/programs/encounter/insert.rs:54]: not sure what you mean, could you please give some more details?

**Comment by clemens-msupply** [comment]
> Thank you for replies and changes so far.
> 
>     1. I think we can use enrolment table for `validate_patient_program_exists` and remove parent_id from document_registry

Yes that works, I did a commit but leave removing of the parent_id for another PR, there is some FE involved as well and I would like this PR merged

> 
>     2. I also think that program for enrolment and other tables was clearer. And I think it should be document_context, since it explicit in the meaning, and other columns that denormalised from document have `document_` prefix (in the future I would assume it would be changed to program_id, and program would then link to context, i.e. encounter ->  program_id -> context -> permission)

not sure what you mean about the table name, what about the new `program` requisition table? (btw. we could use the new `program` table instead of the registry table.)  Lets discuss how to use the context table in the office....


> When say encounter configuration is updated (new field is added) would you create a new document_registry or keep existing one ?

The idea is to have the registry as a reference when creating new documents. For example, which program_enrolment types there are, which encounters belong to each enrolment and which schema there are. When a schema changes a new reference to the new schema is updated in the document registry. So old documents can still refer the old schema they used. New documents can use the most up-to-date schema as specified in the registry.

> 
> > For type and context it's more convenience, these columns shouldn't change for an existing document...
> 
> I think we need an issue to normalise and generally avoid denormalisation (unless it's absolutely necessary, like when we need to snapshot state of related record since it may change), it maybe convenient when it's originally written (when it's fully understood), but duplicates can create confusion when someone new is looking at the code or when looking at the code in the future.

Yes its kind of a snapshot here. The document version is immutable and all the information that was used when creating it is stored in it. For example, the document type is also used to calculate the document hash and then its good to have all data in one place...

> 
> > For example, if a patient is enrolled in the same program at two different clinics there should only be one enrolment
> 
> How would this work in practice ? Since I would assume system would not allow you to enrol patient more then once and all patient related data would sync to sites where they 'exist', would it be if patient 'exists' in both sites and one sites does not have internet while patient visits one site and enrols in a program then visit the site with no internet and enrols there ?

Yes a patient has to visit two sites without the sites syncing in the meantime. After the sync the data from the latest enrolment is used.

**Comment by andreievg** [comment]
>2. I also think that program for enrolment and other tables was clearer. And I think it should be document_context, since it explicit in the meaning, and other columns that denormalised from document have `document_` prefix (in the future I would assume it would be changed to program_id, and program would then link to context, i.e. encounter ->  program_id -> context -> permission)
>> not sure what you mean about the table name, what about the new program requisition table? (btw. we could use the new program table instead of the registry table.) Lets discuss how to use the context table in the office....

I meant that `context` field on enrolment, encounter and program_event is very `vague`. I kind of liked the pervious `program` because enrolment, encounter and program_event can only link to a programs (there is no other context types apply to). For now thought document_context works better then context.

**Review by andreievg — APPROVED** [review]
Thanks for changes and explanation.

After further discussion, approving, looks solid and sounds like the concern over context will be resolved with schema changes. 

Something like this (although i think context table was considered, it looked like it would only have 'id' for now).

```mermaid
erDiagram
    program {
        string id
        string name
        string context_id   
    }
    user_permissions }o--|| program : context_id
    user_permissions {
        string id
        string store_id
        string permission
        string context_id
    }
    user_permissions }o--|| document : context_id
    document {
        string id
        string context_id
    }
    program }o--|| encounter : program_id
    encounter ||--|| document : document_id
    encounter {
        string id
        string document_id
        string program_id
    }
    program }o--|| enrolment : program_id
    enrolment ||--|| document : document_id
    enrolment {
        string id
        string document_id
        string program_id
    }
```

**Pull request merged**